app/src/main/java/ch/hikemate/app/model/route/HikesViewModel.kt

```diff
@@ -0,0 +1,1166 @@
+package ch.hikemate.app.model.route
+
+import android.util.Log
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import ch.hikemate.app.model.elevation.ElevationService
+import ch.hikemate.app.model.route.saved.SavedHike
+import ch.hikemate.app.model.route.saved.SavedHikesRepository
+import ch.hikemate.app.utils.RouteUtils
+import com.google.firebase.Timestamp
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+import org.osmdroid.util.BoundingBox
+
+/**
+ * View model to work with hikes.
+ *
+ * This view model is responsible for a central part of the app's functionality, it handles hikes.
+ *
+ * @param savedHikesRepo The repository to work with saved hikes.
+ * @param osmHikesRepo The repository to work with hikes from OpenStreetMap.
+ * @param elevationRepo The service to retrieve elevation data.
+ * @param dispatcher The dispatcher to be used to launch coroutines.
+ */
+class HikesViewModel(
+    private val savedHikesRepo: SavedHikesRepository,
+    private val osmHikesRepo: HikeRoutesRepository,
+    private val elevationRepo: ElevationService,
+    private val dispatcher: CoroutineDispatcher = Dispatchers.IO
+) : ViewModel() {
+  companion object {
+    private const val LOG_TAG = "HikesViewModel"
+  }
+
+  private val _hikesMutex = Mutex()
+
+  private val _savedHikesMap = mutableMapOf<String, SavedHike>()
+
+  private var _hikeFlowsMap = mutableMapOf<String, MutableStateFlow<Hike>>()
+
+  private val _loading = MutableStateFlow(false)
+
+  private val _hikeFlowsList = MutableStateFlow<List<StateFlow<Hike>>>(emptyList())
+
+  private fun updateHikeFlowsList() {
+    _hikeFlowsList.value = _hikeFlowsMap.values.toList()
+  }
+
+  private var _selectedHikeId: String? = null
+  private val _selectedHike = MutableStateFlow<Hike?>(null)
+
+  /**
+   * Helper function to update the selected hike once (or before) [hikeFlows] has been updated.
+   *
+   * This function does not acquire the [_hikesMutex]. It is the responsibility of the caller to
+   * call this function inside of a [Mutex.withLock] block.
+   *
+   * This function does not switch context either, it is the responsibility of the caller to call
+   * this function inside of a [withContext] block.
+   */
+  private fun updateSelectedHike() {
+    // If there is no selected hike, don't bother.
+    if (_selectedHikeId == null) {
+      return
+    }
+
+    // Retrieve the selected hike from the map. If it's null, don't bother.
+    val selectedHike = _selectedHike.value ?: return
```

Comment:
I assume you mean this :

```kotlin
// If there is no selected hike, don't bother.
if (_selectedHikeId == null || _selectedHike.value == null) {
  return
}

// Retrieve the selected hike from the map.
val selectedHike = _selectedHike.value
```

The reason why I used the Elvis operator is that this way `selectedHike` has type `Hike` (cannot be `null`).

With the method above, `selectedHike` has type `Hike?`, because Kotlin assumes `_selectedHike.value` might have changed between the check and setting the value.

I committed an in-between version where the value is graved in stone *before* the check :

```kotlin
// If there is no selected hike, don't bother.
val selectedHike = _selectedHike.value
if (_selectedHikeId == null || selectedHike == null) {
  return
}
```

This way the compiler understands `selectedHike` is not `null` once the check is passed.

Reply:
Well actually now that I work on your next comment (removing `_selectedHikeId`), the condition become too simple for the compiler, and it tells me to replace with an Elvis operator 😂

Reply:
If the compiler says it 😂 My idea was to have something like this:
```kt
// If there is no selected hike, don't bother.
if (selectedHike.value == null) {
  return
}

// Retrieve the selected hike from the map.
val selectedHike = _selectedHike.value!!